## ひろしまQuest2020#stayhome【コース予測部門】
#### all_merge_14
#### all_merge_13

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_14.f'
SUBMIT = 'submit/course_14.csv'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 168)

### 不要な列を削除

In [4]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [5]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [6]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 153)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  bc_straight  \
0           0           0        0         0           0     0.477023   
1           0           0        0         0           1     0.459832   
2           0           0        0         0           2     0.469730   
3           0           0        1         0           0     0.477023   
4           0           0        1         0           1     0.459832   

   bc_curve  bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  \
0  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
1  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   
2  0.066062   0.159395  0.031361  0.162940     0.062722   0.014112   
3  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
4  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   

   bc_cutball  bc_area0  bc_area1  bc_area2  bc_area3  bc_area4  bc_area5  \
0    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
1    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   
2    0.033679  0.021884  0.035656  0.030884  0.020316  0.021748  0.021816   
3    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
4    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   

   bc_area6  bc_area7  bc_area8  bc_area9  bc_area10  bc_area11  bc_area12  \
0  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
1  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   
2  0.028429  0.036678  0.029520  0.133761   0.131170   0.201595   0.286542   
3  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
4  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   

   first  second  third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0       0      0         0       0.0     1.0        1     1        0   
1      0       0      0         0       0.0     1.0        1     1        0   
2      0       0      0         0       0.0     1.0        1     1        0   
3      0       0      0         0       0.0     0.0        1     1        0   
4      0       0      0         0       0.0     0.0        1     1        0   

   date_from_opening  ...  weight_bat  draft_order_bat  salary_bat  \
0                  0  ...          74              5.0       15000   
1                  0  ...          74              5.0       15000   
2                  0  ...          74              5.0       15000   
3                  0  ...          74              1.0        7040   
4                  0  ...          74              1.0        7040   

   f

### train
- 行数: 257117

In [7]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 153)

### test
- 行数: 521650

In [8]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 153)

In [9]:
del all_pitch
gc.collect()

60

In [10]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [11]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [12]:
lgb_train = lgb.Dataset(train_d, train['pitch_area'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [13]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 13,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 8.769293390201968, 
        'lambda_l2': 3.913949617576324e-05, 
        'num_leaves': 6, 
        'feature_fraction': 0.4, 
        'bagging_fraction': 0.8391111798378441, 
        'bagging_freq': 4, 
        'min_child_samples': 50
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 2.35314 + 0.000937687
[200]	cv_agg's multi_logloss: 2.34668 + 0.0010113
[300]	cv_agg's multi_logloss: 2.34408 + 0.00102303
[400]	cv_agg's multi_logloss: 2.34258 + 0.000991686
[500]	cv_agg's multi_logloss: 2.3415 + 0.00108701
[600]	cv_agg's multi_logloss: 2.34089 + 0.00118182
[700]	cv_agg's multi_logloss: 2.34047 + 0.00120034
[800]	cv_agg's multi_logloss: 2.34018 + 0.00117776
[900]	cv_agg's multi_logloss: 2.34009 + 0.00111678
[1000]	cv_agg's multi_logloss: 2.34011 + 0.00111843
Best num_boost_round: 909
Best CV score: 2.340082741044867
CPU times: user 1h 16min 55s, sys: 15.3 s, total: 1h 17min 10s
Wall time: 9min 40s


In [14]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [15]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 23min 32s, sys: 4.42 s, total: 23min 36s
Wall time: 2min 57s


### Feature Importance

In [16]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
100             weight_bat       826
1        pitch_cnt_in_game       828
95            pit_game_cnt       829
84         salary_year_pit       837
99              height_bat       840
144  salary_x_year_dif_b-c       868
8    pitcher_cnt_in_inning       872
0               No_in_game       882
140         salary_dif_b-c       885
86                 bmi_pit       898
91                changeup       905
137    salary_year_dif_p-c       928
113           bat_game_cnt       930
111                bmi_bat       962
47             elapsed_min       965
131    salary_year_dif_p-b       973
94                   total       979
89                   shoot       979
143    salary_year_dif_b-c      1024
97          pit_batter_cnt      1027
7      pitcher_cnt_in_game      1037
139            bmi_dif_p-c      1078
87                   curve      1084
90                    fork      1095
112             batter_cnt      1107
133            bmi_dif_p-b      1179
145            bmi_dif_b-c      1180
88                  slider      1364
46       date_from_opening      1961
48                min_diff      2011

In [17]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 6min 6s, sys: 588 ms, total: 6min 7s
Wall time: 47 s


## 結果出力

In [18]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.052407  0.142541  0.060116  0.042920  0.070804  0.033387   
1      1  0.047790  0.144078  0.079380  0.036273  0.058967  0.038422   
2      2  0.027748  0.083917  0.057449  0.023302  0.035546  0.030192   
3      3  0.034463  0.099420  0.064195  0.030023  0.042649  0.025761   
4      4  0.038971  0.115826  0.064219  0.036125  0.043660  0.040346   

          6         7         8         9        10        11        12  
0  0.055867  0.055533  0.033525  0.187610  0.091477  0.086205  0.087608  
1  0.041201  0.065164  0.038256  0.163279  0.073721  0.098737  0.114732  
2  0.039095  0.048078  0.028609  0.178813  0.090817  0.212279  0.144156  
3  0.048872  0.060133  0.035961  0.198695  0.096317  0.126009  0.137502  
4  0.055306  0.059018  0.041634  0.175699  0.077472  0.120763  0.130960

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [19]:
submit.mean()

index    260824.500000
0             0.048956
1             0.069453
2             0.055621
3             0.046908
4             0.060590
5             0.046583
6             0.043097
7             0.064742
8             0.052816
9             0.111640
10            0.088200
11            0.133273
12            0.178122
dtype: float64

In [20]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [21]:
'signate submit --competition-id=276 ~/work/baseball/{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=276 ~/work/baseball/submit/course_14.csv --note feat=153_cv=2.340082741044867'

### GCS

In [22]:
'gsutil cp ~/work/baseball/{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ~/work/baseball/submit/course_14.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648